### PR TITLE
create istio-ca-root-cert cm for each remote ns

### PR DESCRIPTION
**Pull request by irisdingbj**
https://github.com/istio/istio/issues/23138

1. `--set values.global.centralIstiod=true ` 
2.  central istiod will create istio-ca-root-cert cm for each remote ns

- default install
- central istiod **will not** create istio-ca-root-cert cm remote ns

manually tested above two scenarios and passed

### CODE REVIEW

**Review by GregHanson — APPROVED** [review]

**Review by linsun — APPROVED** [review]
LG, thank you @irisdingbj !

**Review by lei-tang — CHANGES_REQUESTED** [review]

File: manifests/charts/base/templates/clusterrole.yaml
Comment (howardjohn) [inline manifests/charts/base/templates/clusterrole.yaml:132]: nit: having a "istio-reader" role with write access is a bit misleading. Although it already has tokenreviews write access..

File: manifests/charts/base/templates/clusterrole.yaml
Comment (costinm) [inline manifests/charts/base/templates/clusterrole.yaml:130]: Please add resourceName as well - we don't want to read all the config maps in all namespaces, just the ones with our fixed name.

**Comment by lei-tang** [comment]
@irisdingbj To the PR description or the PR code, can you add a description of the workflow involved with this PR? E.g., where is the root CA certificate obtained? what is a remote ns? how to configure a ns as a remote ns?

File: pilot/pkg/serviceregistry/kube/controller/multicluster.go
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/multicluster.go:119]: This won't support the root cert changing like the non-multicluster variant does. Is that ok?

File: pilot/pkg/serviceregistry/kube/controller/multicluster.go
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/multicluster.go:113]: is this function only called when the member cluster doesn't have an istiod itself?

**Review by howardjohn — COMMENTED** [review]
We should not consider this a supported feature until it has proper e2e testing of this topology. Not against merging without that, but we should not advertise it

**Review by howardjohn — COMMENTED** [review]
everything in here seems mostly fine but I really have no idea if this works given we have no tests at all

**Comment by lei-tang** [comment]
I agree with John on the importance of the tests. Both unit tests and integration tests should be added to cover the feature related to this PR. The unit tests can be added to this PR and if the implementation of the feature is incomplete yet, the PR should add a to-do comment for the integration tests so that the integration tests will be addressed by follow-up PRs.

File: pilot/pkg/serviceregistry/kube/controller/multicluster.go
Comment (costinm) [inline pilot/pkg/serviceregistry/kube/controller/multicluster.go:55]: We have Options object with all kind of options - can we put the caroot there as well ?

**Review by costinm — COMMENTED** [review]
Great work ! 

Few cosmetic changes.

**Comment by costinm** [comment]
Re. testing: single cluster behavior is covered already by e2e tests. 
MC behavior is also covered, at least in the common case.

I think we should have a e2e test for the entire 'central Istio' use case - but for 1.6 since it's new 
we can probably test manually.

What is missing and must be fixed: an env or mesh config or annotation on the MC secret to turn on or off this feature. I'm pretty sure it'll have really bad effects in some configurations where you have multiple istiod clusters, and each will attempt to create the config maps - possibly with different roots.
And that's hard to capture in tests.

So please make it 'opt in' and off by default, with the option to enable including enough info for users to know to not turn it on unless they are in this scenario.

File: pilot/pkg/serviceregistry/kube/controller/multicluster.go
Comment (costinm) [reply pilot/pkg/serviceregistry/kube/controller/multicluster.go:119]: I think it expires in 10 years or so - and a restart of istiod in the next 10 years will handle this.

**Comment by howardjohn** [comment]
> So please make it 'opt in' and off by default

+1, that is my main concern that it may impact multiple istiod clusters.
perfectly fine with approving if its disabled

File: manifests/charts/base/templates/clusterrole.yaml
Comment (costinm) [reply manifests/charts/base/templates/clusterrole.yaml:132]: +1 - using a new name would also be good for allowing both use cases, the 'central cluster in control' is very different than 'each cluster running istiod and just reading endpoints'. 

I understand this may require changes in docs/istioctl/etc to use the right SA for the case - but this change breaks the fundamental security assumptions of MC 'only reads endpoints'.

For tokenreview - I haven't reviewed that, but same security considerations should be used. 
If tokenreview 'write' is only used to validate - it might be equivalent with a read (with a horrible API). If it can be for anything else - then separate role and SA should be used.

**Comment by costinm** [comment]
On John's comment: I hope nobody is expecting this to be a 'supported feature' in 1.6, it just landed. 
It is alpha or experimental - and once enough testing and feedback is available we can move to beta, etc.

**Comment by linsun** [comment]
Thank you for the valuable comment @howardjohn @costinm @lei-tang !  I agree with opt-in for this.

File: pilot/pkg/serviceregistry/kube/controller/multicluster.go
Comment (linsun) [reply pilot/pkg/serviceregistry/kube/controller/multicluster.go:113]: I think it will regardless, so we can use the feature flag to guard this for now.

File: pilot/pkg/serviceregistry/kube/controller/multicluster.go
Comment (irisdingbj) [reply pilot/pkg/serviceregistry/kube/controller/multicluster.go:55]: put it into Options

File: pilot/pkg/bootstrap/multicluster.go
Comment (irisdingbj) [reply pilot/pkg/bootstrap/multicluster.go:25]: thanks! removed this args and make it part of opts according to @costinm 's suggestion.

File: manifests/charts/base/templates/clusterrole.yaml
Comment (irisdingbj) [reply manifests/charts/base/templates/clusterrole.yaml:132]: add resourcename to restrict the usage

File: pilot/pkg/serviceregistry/kube/controller/multicluster.go
Comment (irisdingbj) [reply pilot/pkg/serviceregistry/kube/controller/multicluster.go:113]: guard by `values.global.centralIstiod=false` and default is off

**Comment by irisdingbj** [comment]
@linsun @howardjohn @costinm @lei-tang @GregHanson Thanks for the review, PTAL.

File: manifests/charts/base/templates/clusterrole.yaml
Comment (irisdingbj) [reply manifests/charts/base/templates/clusterrole.yaml:131]: ```Work item handle failed (error when creating configmap istio-ca-root-cert: configmaps is forbidden: User "system:serviceaccount:istio-system:istio-reader-service-account" cannot create resource "configmaps" in API group "" in the namespace "uu"), retry after delay 1s```
if I added resourceNames here, it does not work. 
so I changed back and let it guard by our global flag

**Review by lei-tang — APPROVED** [review]

File: pilot/pkg/serviceregistry/kube/controller/controller.go
Comment (lei-tang) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:117]: Suggest changing all CAROOT to CARoot to be consistent with the variable naming convention.

File: pilot/pkg/serviceregistry/kube/controller/multicluster.go
Comment (lei-tang) [inline pilot/pkg/serviceregistry/kube/controller/multicluster.go:42]: Suggest changing all caroot to caRoot to be consistent with the variable naming convention.

**Review by lei-tang — COMMENTED** [review]
Approved to unblock the PR. Meanwile, please add unit tests if you can,  and fix the variable namings (e.g., CAROOT, caroot, etc).

File: pilot/pkg/serviceregistry/kube/controller/multicluster.go
Comment (linsun) [reply pilot/pkg/serviceregistry/kube/controller/multicluster.go:113]: Thanks!

**Comment by linsun** [comment]
@irisdingbj thank you for adding the opt in flag by user's choice.  @howardjohn @costinm I think this is safe to merge, can you check?

The additional feedback that should be addressed in a separate PR:
- caRoot comment from @lei-tang 
- possible unit test if we can
- add an additional writer service account only for ConfigMap, and only for the ConfigMap called istio-ca-root-cert
- investigate and test impact with multi-revision support. We are currently under the assumption that multi-revision would share the same root cert so the impact may be minimal or none.

Let me know anything else I missed.

File: pilot/pkg/bootstrap/server.go
Comment (howardjohn) [inline pilot/pkg/bootstrap/server.go:260]: nit: if you want to get root cert rotation for "free" just change CAROOT to be `func() string` instead. Or `func() map[string]string` which is the input to the namespace controller. That is what the local one does

**Review by howardjohn — APPROVED** [review]
lgtm

**Comment by irisdingbj** [comment]
> @irisdingbj thank you for adding the opt in flag by user's choice. @howardjohn @costinm I think this is safe to merge, can you check?
> 
> The additional feedback that should be addressed in a separate PR:
> 
> * caRoot comment from @lei-tang
> * possible unit test if we can
> * add an additional writer service account only for ConfigMap, and only for the ConfigMap called istio-ca-root-cert
> * investigate and test impact with multi-revision support. We are currently under the assumption that multi-revision would share the same root cert so the impact may be minimal or none.
> 
> Let me know anything else I missed.

Thanks Lin for this summary, opened an issue https://github.com/istio/istio/issues/23263 to track

**Comment by nmittler** [comment]
@irisdingbj I believe this breaks my cluster-local change from https://github.com/istio/istio/pull/23150/files#diff-8cc1e432737fa0d93dc76b5bc10cac83R159

You removed the setting of the `DomainSuffix` on the environment. Was this intentional?

**Comment by irisdingbj** [comment]
@nmittler Thanks for taking care of this.  My fault for bad merge :(

**Comment by linsun** [comment]
@irisdingbj I think this change is reverted from master but still exist in 1.6.  Could you double check?  We still want the function in master as well.

**Comment by irisdingbj** [comment]
@linsun change is not reverted in master. But need new change since https://github.com/istio/istio/pull/23684 changed the ca certs behaviour.

**Pull request merged**
